src/Datasource/Paginator.php

```diff
@@ -265,26 +263,31 @@ protected function buildParams(array $data)
             $directionDefault = current($defaults['order']);
         }
 
-        $start = 0;
-        if ($count >= 1) {
-            $start = (($page - 1) * $limit) + 1;
+        $pageCount = 0;
+        if ($count !== null) {
+            $pageCount = max((int)ceil($count / $limit), 1);
+            $page = min($page, $pageCount);
+        } elseif ($numResults === 0 && $requestedPage > 1) {
+            $page = 1;
         }
-        $end = $start + $limit - 1;
-        if ($count < $end) {
-            $end = $count;
+
+        $start = $end = 0;
+        if ($numResults > 0) {
+            $start = (($page - 1) * $limit) + 1;
+            $end = $start + $numResults - 1;
         }
 
         $paging = [
             'finder' => $data['finder'],
             'requestedPage' => $requestedPage,
             'page' => $page,
-            'current' => $data['numResults'],
+            'current' => $numResults,
             'count' => $count,
             'perPage' => $limit,
             'start' => $start,
             'end' => $end,
             'prevPage' => $page > 1,
-            'nextPage' => $count > ($page * $limit),
+            'nextPage' => $count === null ? true : ($count > ($page * $limit)),
```

Comment:
Should determining the next/prev pages be moved into methods that implementations can override? Same for start/end?

Reply:
Okay, with do.